training_pipeline/models/model.py

```diff
@@ -1,266 +1,287 @@
-import datetime
-import os
-from typing import List
+from typing import List, Dict, Tuple
+
 import absl
-import keras_tuner
 import tensorflow as tf
 from tensorflow.keras.optimizers import Adam
-import tensorflow_transform as tft
-
-from tensorflow_cloud import CloudTuner
-from tfx.v1.components import TunerFnResult
-from tfx.components.trainer.fn_args_utils import DataAccessor
 from tfx.components.trainer.fn_args_utils import FnArgs
-from tfx.dsl.io import fileio
-from tfx_bsl.tfxio import dataset_options
-import tfx.extensions.google_cloud_ai_platform.constants as vertex_const
-import tfx.extensions.google_cloud_ai_platform.trainer.executor as vertex_training_const
-import tfx.extensions.google_cloud_ai_platform.tuner.executor as vertex_tuner_const
 
-_TRAIN_DATA_SIZE = 128
-_EVAL_DATA_SIZE = 128
-_TRAIN_BATCH_SIZE = 32
-_EVAL_BATCH_SIZE = 32
-_CLASSIFIER_LEARNING_RATE = 1e-3
-_FINETUNE_LEARNING_RATE = 7e-6
-_CLASSIFIER_EPOCHS = 30
-
-_IMAGE_KEY = "image"
-_LABEL_KEY = "label"
+_CONCRETE_INPUT = "pixel_values"
+_IMAGE_SHAPE = (128, 128)
+_TRAIN_BATCH_SIZE = 64
+_EVAL_BATCH_SIZE = 64
+_EPOCHS = 2
+_LR = 0.00006
 
 
 def INFO(text: str):
     absl.logging.info(text)
 
 
-def _transformed_name(key: str) -> str:
-    return key + "_xf"
+"""
+    _serving_preprocess, _serving_preprocess_fn, and 
+    _model_exporter functions are defined to provide pre-
+    processing capabilities when the model is served.
+"""
 
 
-def _get_signature(model):
-    signatures = {
-        "serving_default": _get_serve_image_fn(model).get_concrete_function(
-            tf.TensorSpec(
-                shape=[None, 224, 224, 3],
-                dtype=tf.float32,
-                name=_transformed_name(_IMAGE_KEY),
-            )
-        )
+def _serving_preprocess(string_input):
+    decoded_input = tf.io.decode_base64(string_input)
+    decoded = tf.io.decode_jpeg(decoded_input, channels=3)
+    decoded = decoded / 255
```

Comment:
The preprocessing utilities are slightly different from what was used in the TFRecord creation script.

Reply:
meaning first `normalize` then `resize`?

Reply:
You missed out subtracting -1 from the pixel values. Please take a look again.

Reply:
hm `label -= 1` is for label data. But, since this is a preprocessing function for the serving, we don't have label data. Or do I miss something?

Reply:
Oh, then you're right. It's very hard to review 21 changed files in a single PR honestly :D